Add BGE-M3 encode test

gengelageTT opened the pull request:
### Ticket
#1196 

### Problem description
We want bge_m3 encode running in CI and recorded by superset. The current TorchModelTester doesn't support inputs/outputs that aren't tensors and so this model can't use the existing test infra.

### What's changed
Added a unique test for BAAI/bge-m3 encode_single_device. This test loads the model from tt-forge-models, runs on cpu and device then computes the PCC. PCC is currently ~~0.99~~ [EDIT]: _0.98 on bh_ and asserted on so we can catch regressions. This test will run in the nightly CI and report to superset.

Why TorchModelTester is not used: 
- The inputs are sentences which cannot be moved onto device using `.to(device)`. In this test the inputs are moved onto device after tokenization by passing in the device as an argument to `encode_single_device`. This is why we use `encode_single_device` instead of `encode` (encode doesn't take in a device argument)
- Some outputs aren't torch tensors and calls like `torch.allclose()` fail to run when comparing them.
- `model.compile` doesn't support compiling `methods` but `torch.compile` does

### Checklist
- [x] New/Existing tests provide coverage for changes

gengelageTT commented:
Re-running test in CI here: https://github.com/tenstorrent/tt-xla/actions/runs/17990885856

AleksKnezevic commented on tests/torch/single_chip/models/bge_m3/test_bge_m3_encode.py:
```diff
@@ -0,0 +1,194 @@
+MODEL_INFO = ModelLoader.get_model_info(VARIANT_NAME)
+
+
+def calculate_pcc(tensor1, tensor2):
```
Is it possible to reuse the calculation from: `torch_comparator.py'

gengelageTT replied:
but I can try copy the calculation from that pr if we want it to match what's done in the test infra a closely as possible

AleksKnezevic replied:
Yeah, I think it's worthwhile adding that.

AleksKnezevic reviewed (approved):
If possible, please use pcc calculation from torch_comparator, otherwise looks good.

gengelageTT commented:
pcc lowered to 0.92 which means bringup status is now `INCORRECT_RESULT`. Test is passing in CI here: https://github.com/tenstorrent/tt-xla/actions/runs/17991553896/job/51182748319
Added issue to investigate lower pcc on bh here: https://github.com/tenstorrent/tt-xla/issues/1461

gengelageTT commented on tests/torch/single_chip/models/bge_m3/test_bge_m3_encode.py:
```diff
@@ -0,0 +1,194 @@
+    # Connect the device.
+    device = xm.xla_device()
+
+    # Move inputs and model to device.
```
I'll change this to run the not-compiled model for cpu. Normally we would move the model to device, but the `encode_single_device` function handles this internally when we pass in the device as an argument: https://github.com/FlagOpen/FlagEmbedding/blob/master/FlagEmbedding/inference/embedder/encoder_only/m3.py#L340

gengelageTT commented:
@AleksKnezevic I adjusted pcc calculation to more closely follow what's done in the TorchComparator. This required adding the following 2 functions:
-  `tree_map_to_torch()` turns the lexical weights dictionary into a tensor of values so that the pcc compares the outputted tensors instead of the individual items (individual items -> nan pcc). 
- `correct_golden_lexical_weights()`fixes shape mismatches between the output. Since `encode_single_device` [removes lexical weights that are 0](https://github.com/FlagOpen/FlagEmbedding/blob/master/FlagEmbedding/inference/embedder/encoder_only/m3.py#L358) (after the computation done on device) blackhole tests were failing since some values were incorrectly non-zero, leading to [shape mismatches](https://github.com/tenstorrent/tt-xla/actions/runs/18013919884/job/51253819350#step:15:181). 

@mrakitaTT I also moved `FlagEmbedding` to `venv/requirements-dev.txt` and created an issue to add support for methods/functions and non-tensor inputs: https://github.com/tenstorrent/tt-xla/issues/1471

changes are passing in CI here: https://github.com/tenstorrent/tt-xla/actions/runs/18019902210

kmabeeTT reviewed (approved):
Awesome! Let's ship it.

AleksKnezevic commented:
I agree!

mrakitaTT commented:
+1 :)
